### PR TITLE
GSK-1096 - GSK-1084 - No code slicing function

### PR DESCRIPTION
**Pull request by kevinmessiaen**
## Description

- [x] No code filter from the UI
- [x] Load no code filter in the ml worker 

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [x] 🥂 Improvement (non-breaking change which improves an existing feature)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CONTRIBUTING.md) guide.
- [ ] I've updated the code style using `make codestyle`.
- [ ] I've written tests for all new methods and classes that I created.
- [ ] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

**Comment by kevinmessiaen** [comment]
GSK-1084

File: frontend/src/views/main/project/modals/CreateSliceModal.vue [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+const props = defineProps<{
+    dataset: DatasetDTO
+}>();
+
+const columnFilters = ref<Array<Partial<NoCodeFilter>>>([{}])
+
+const emit = defineEmits(['created']);
+
+const CLAUSE_BUILDERS: { [clause in NoCodeSlicingType]: (column: string, value: string) => string } = {
+    [NoCodeSlicingType.IS]: (column, value) => `giskard.slicing.slice.EqualTo(${column}, ${value})`,
```
Comment (andreybavt) [inline frontend/src/views/main/project/modals/CreateSliceModal.vue:52]: I don't think UI should be aware of the python implementation details. Doing it on the backend is better

File: frontend/src/views/main/project/modals/CreateSliceModal.vue [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+function clauseToString(clause: NoCodeFilter) {
+    return `${escapePython(clause.column)} ${CLAUSE_SYMBOLS[clause.slicingType]} ${NO_VALUES.includes(clause.slicingType) ? '' : escapePython(clause.value, clause.column)}`
+}
+
+async function save(close) {
+    const name = columnFilters.value.map((v) => clauseToString(v as NoCodeFilter)).join(' & ')
+
+    const dto: SlicingFunctionDTO = {
+        uuid: uuidv4(),
```
Comment (andreybavt) [inline frontend/src/views/main/project/modals/CreateSliceModal.vue:108]: along the line with the previous comment, UI should be sending what the user provides (`columnFilters` , `name`), the rest should be done on the backend

File: frontend/src/views/main/utils/SlicingFunctionSelector.vue [hunk 3]
```diff
@@ -14,7 +14,19 @@
             :dense="fullWidth"
             hide-details
             :prepend-inner-icon="icon ? 'mdi-knife' : null"
-        ></v-select>
+        >
+            <template v-slot:append-item v-if="allowNoCodeSlicing">
+                <!-- TODO add NO CODE slices -->
+                <v-list-item @click="createSlice">
```
Comment (andreybavt) [inline frontend/src/views/main/utils/SlicingFunctionSelector.vue:20]: anything else to do for this TODO?

File: frontend/package.json [hunk 4]
```diff
@@ -33,6 +33,7 @@
     "postcss": "^8",
     "register-service-worker": "^1.7.2",
     "rrweb-snapshot": "^1.1.14",
+    "uuid": "^9.0.0",
```
Comment (andreybavt) [inline frontend/package.json:36]: no need for exra dependency in case we do slice creation on the backend using conditions sent from the frontend

**Review by andreybavt — CHANGES_REQUESTED** [review]

**Comment by andreybavt** [comment]
@kevinmessiaen , could you also have a look at Sonar issues in this PR?

File: backend/src/main/java/ai/giskard/web/dto/ComparisonType.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package ai.giskard.web.dto;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+@Getter
+public enum ComparisonType {
+    IS(true, "==", "giskard.slicing.slice.EqualTo(%s, %s)"),
```
Comment (andreybavt) [inline backend/src/main/java/ai/giskard/web/dto/ComparisonType.java:9]: this is going to be hard to maintain if the backend becomes aware of python implementation details. I think we should just pass a list of rules (language agnostic) from the front to python and then python would know how to convert for example "IS_NOT" to `NotEqualTo`

File: backend/src/main/java/ai/giskard/service/SlicingFunctionService.java [hunk 6]
```diff
@@ -36,4 +46,46 @@ protected SlicingFunction create(SlicingFunctionDTO dto) {
...
+        } else {
+            return String.format("%s %s", escapePythonVariable(comparisonClause.getColumnName()), comparisonType.getSymbol());
+        }
+    }
+
+    public SlicingFunctionDTO generate(List<ComparisonClauseDTO> comparisonClauses) {
+        String name = comparisonClauses.stream().map(this::clauseToString).collect(Collectors.joining(" & "));
+
+        SlicingFunction slicingFunction = new SlicingFunction();
+        slicingFunction.setUuid(UUID.randomUUID());
```
Comment (andreybavt) [inline backend/src/main/java/ai/giskard/service/SlicingFunctionService.java:74]: should we be using uuid5 to generate the same uuid for the identical rules?

**Review by andreybavt — CHANGES_REQUESTED** [review]

File: backend/src/main/resources/config/liquibase/changelog/20230531185949_changelog.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    <changeSet author="kevinmessiaen (generated)" id="1685552392648-2">
+        <dropDefaultValue columnDataType="varchar(50)" columnName="created_by" tableName="inspections"/>
+    </changeSet>
+    <changeSet author="kevinmessiaen (generated)" id="1685552392648-3">
+        <dropDefaultValue columnDataType="int" columnName="process_type" tableName="callable_functions"/>
+    </changeSet>
+    <changeSet author="kevinmessiaen (generated)" id="1685552392648-4">
+        <dropDefaultValue columnDataType="boolean" columnName="sample" tableName="inspections"/>
+    </changeSet>
+</databaseChangeLog>
```
Comment (Googleton) [inline backend/src/main/resources/config/liquibase/changelog/20230531185949_changelog.xml:17]: Is this one all good ? I don't see changes in the code to the inspections table, but see some here?

Comment (Googleton) [reply backend/src/main/resources/config/liquibase/changelog/20230531185949_changelog.xml:17]: Actually it's not even in the master.xml

Comment (kevinmessiaen) [reply backend/src/main/resources/config/liquibase/changelog/20230531185949_changelog.xml]: Yes you're right I comited it by mistake, I removed it

File: python-client/giskard/slicing/slice.py [hunk 8]
```diff
@@ -69,13 +160,73 @@ def __str__(self) -> str:
...
+
+def generate_clause(clause: Dict[str, str]) -> Clause:
+    if clause['comparisonType'] == 'IS':
+        return EqualTo(clause['columnName'], clause['value'])
+    elif clause['comparisonType'] == 'IS_NOT':
+        return NotEqualTo(clause['columnName'], clause['value'])
+    elif clause['comparisonType'] == 'CONTAINS':
+        return ContainsWord(clause['columnName'], clause['value'], is_not=False)
+    elif clause['comparisonType'] == 'DOES_NOT_CONTAINS':
+        return ContainsWord(clause['columnName'], clause['value'], is_not=True)
```
Comment (Googleton) [inline python-client/giskard/slicing/slice.py:210]: Why have is_not ? Could just return ContainsWord or !ContainsWord ?

File: python-client/giskard/slicing/slice.py [hunk 9]
```diff
@@ -69,13 +160,73 @@ def __str__(self) -> str:
...
+    elif clause['comparisonType'] == 'STARTS_WITH':
+        return StartsWith(clause['columnName'], clause['value'])
+    elif clause['comparisonType'] == 'ENDS_WITH':
+        return EndsWith(clause['columnName'], clause['value'])
+    elif clause['comparisonType'] == 'GREATER_THAN_EQUALS':
+        return GreaterThan(clause['columnName'], clause['value'], equal=True)
+    elif clause['comparisonType'] == 'GREATER_THAN':
+        return GreaterThan(clause['columnName'], clause['value'])
+    elif clause['comparisonType'] == 'LOWER_THAN_EQUALS':
+        return LowerThan(clause['columnName'], clause['value'], equal=True)
```
Comment (Googleton) [inline python-client/giskard/slicing/slice.py:220]: Same here, why add a custom parameter and not just return LowerThan OR EqualTo ? Would make functions do only what their name says

File: python-client/giskard/slicing/slice.py [hunk 10]
```diff
@@ -69,13 +160,73 @@ def __str__(self) -> str:
...
+    elif clause['comparisonType'] == 'LOWER_THAN':
+        return LowerThan(clause['columnName'], clause['value'])
+    elif clause['comparisonType'] == 'IS_EMPTY':
+        return IsNa(clause['columnName'], is_not=False)
+    elif clause['comparisonType'] == 'IS_NOT_EMPTY':
+        return IsNa(clause['columnName'], is_not=True)
```
Comment (Googleton) [inline python-client/giskard/slicing/slice.py:226]: Same as before, not sure if is_not belongs here?

File: python-client/giskard/slicing/slice.py (hunk 8 above)
Comment (kevinmessiaen) [reply python-client/giskard/slicing/slice.py:210]: No because ContainsWord is a `Clause` object that is used on Dataset to filter them

**Pull request merged**